### PR TITLE
fix: (2.38) tracker data view deserialisation issues [DHIS2-12249]

### PR DESCRIPTION
**Pull request by zubaira**
Validations are added based on this document 
https://docs.google.com/document/d/10c8BcARHTMWz-iH2XO0UbR4L57iyUt4OStrmSWbktYU/edit

### CODE REVIEW

**Comment by vilkg** [comment]
Looks like this PR is breaking the contract, so the API tests fail: 
``` {
e2e-test_1  |                 "message": "Property `program` can not be set when property `relationshipEntity` is `PROGRAM_STAGE_INSTANCE`. ",
e2e-test_1  |                 "mainKlass": "org.hisp.dhis.relationship.RelationshipConstraint",
e2e-test_1  |                 "errorCode": "E4023",
e2e-test_1  |                 "errorProperties": [
e2e-test_1  |                     "program",
e2e-test_1  |                     "relationshipEntity",
e2e-test_1  |                     "PROGRAM_STAGE_INSTANCE"
e2e-test_1  |                 ]
e2e-test_1  |             },
e2e-test_1  |             {
e2e-test_1  |                 "message": "Property `programStage` must be set when property `relationshipEntity` is `PROGRAM_STAGE_INSTANCE`. ",
e2e-test_1  |                 "mainKlass": "org.hisp.dhis.relationship.RelationshipConstraint",
e2e-test_1  |                 "errorCode": "E4024",
e2e-test_1  |                 "errorProperties": [
e2e-test_1  |                     "programStage",
e2e-test_1  |                     "relationshipEntity",
e2e-test_1  |                     "PROGRAM_STAGE_INSTANCE"
e2e-test_1  |                 ]
e2e-test_1  |             },```

If this was merged as is, we would need maintenance app changes and a db migration

**Review by vilkg — CHANGES_REQUESTED** [review]
left a comment

**Comment by zubaira** [comment]
> left a comment

uploaded the fix. We will not change the API contract for now.

**Review by enricocolasante — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/RelationshipTypeObjectBundleHook.java [hunk 1]
```diff
@@ -224,18 +195,28 @@ private void validateTrackedEntityInstance( RelationshipConstraint constraint, C
...
 
         TrackedEntityType trackedEntityType = constraint.getTrackedEntityType();
 
+        if ( constraint.getProgramStage() != null )
+        {
+            addReports.accept( new ErrorReport( RelationshipConstraint.class, ErrorCode.E4023, PROGRAM_STAGE,
+                RELATIONSHIP_ENTITY, TRACKED_ENTITY_INSTANCE ) );
+        }
+
+        if ( constraint.getProgram() != null )
```
Comment (Bekkalizer) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/RelationshipTypeObjectBundleHook.java:209]: We need to allow program to be specified for tracked entity types

Comment (zubaira) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/RelationshipTypeObjectBundleHook.java]: resolved

**Review by vilkg — APPROVED** [review]

**Pull request merged**